crates/shared/src/http_client.rs

```diff
@@ -19,11 +23,141 @@ pub async fn response_body_with_size_limit(
+        );
+        Ok(Self {
+            drop_requests_until: Instant::now(),
+            times_rate_limited: 0,
+            back_off_growth_factor,
+            min_back_off,
+            max_back_off,
+        })
+    }
+}
+
+impl RateLimitingStrategy {
+    /// Resets back off and stops rate limiting requests.
+    pub fn response_ok(&mut self) {
+        self.times_rate_limited = 0;
+        self.drop_requests_until = Instant::now();
+    }
+
+    /// Calculates back off based on how often we got rate limited in a row.
+    fn get_current_back_off(&self) -> Duration {
+        let mut back_off = self.min_back_off;
+        for _ in 0..self.times_rate_limited {
+            // do this iteratively to reasonably prevent panics on overflow
+            back_off = back_off.mul_f64(self.back_off_growth_factor);
+            if back_off >= self.max_back_off {
+                return self.max_back_off;
+            }
+        }
+        back_off
+    }
+
+    /// Returns updated back off if no other thread increased it in the mean time.
+    pub fn response_rate_limited(&mut self, previous_rate_limits: u64) -> Option<Duration> {
+        if self.times_rate_limited != previous_rate_limits {
+            // Don't increase back off if somebody else already updated it in the meantime.
+            return None;
+        }
+
+        self.times_rate_limited += 1;
+        let new_back_off = self.get_current_back_off();
+        self.drop_requests_until = Instant::now() + new_back_off;
+        Some(new_back_off)
+    }
+
+    /// Returns number of times we got rate limited in a row if we are currently allowing requests.
+    pub fn times_rate_limited(&self, now: Instant) -> Option<u64> {
+        if self.drop_requests_until > now {
+            return None;
+        }
+
+        Some(self.times_rate_limited)
+    }
+}
+
+#[derive(Debug)]
+pub struct RateLimiter {
+    pub strategy: Mutex<RateLimitingStrategy>,
+}
+
+impl RateLimiter {
+    fn strategy(&self) -> MutexGuard<RateLimitingStrategy> {
+        self.strategy.lock().unwrap()
+    }
+}
+
+impl From<RateLimitingStrategy> for RateLimiter {
+    fn from(strategy: RateLimitingStrategy) -> Self {
+        Self {
+            strategy: Mutex::new(strategy),
+        }
+    }
+}
+
+impl RateLimiter {
+    /// If a request receives the response "Too many requests" (status code 429) future requests
+    /// will get dropped for some time. Every successive 429 response increases that time exponentially.
+    /// When a request eventually returns a normal result again future requests will no longer get
+    /// dropped until the next 429 response occurs.
+    pub async fn request(&self, request: RequestBuilder) -> Result<Response> {
+        let times_rate_limited = match self.strategy().times_rate_limited(Instant::now()) {
+            None => {
+                tracing::warn!("dropping request because API is currently rate limited");
+                anyhow::bail!("rate limited");
```

Comment:
ubernit:
```suggestion
                anyhow::bail!("backing off rate limit");
```

I would use a slightly different error message to distinguish between backing of from rate limiting and being rate limited by the API.

Feel free to ignore this suggestion - I don't feel super strongly about it and we can already distinguish with the additional log messages.